unix/modos.c

```diff
@@ -171,6 +171,36 @@ STATIC mp_obj_t listdir_next(mp_obj_t self_in) {
         return MP_OBJ_STOP_ITERATION;
     }
 
+    return mp_obj_new_str(dirent->d_name, strlen(dirent->d_name), false);;
+}
+
+STATIC mp_obj_t mod_os_listdir(size_t n_args, const mp_obj_t *args) {
+    const char *path = ".";
+    if (n_args > 0) {
+        path = mp_obj_str_get_str(args[0]);
+    }
+    mp_obj_listdir_t *o = m_new_obj(mp_obj_listdir_t);
+    o->base.type = &mp_type_polymorph_iter;
+    o->dir = opendir(path);
+    o->iternext = listdir_next;
+    return MP_OBJ_FROM_PTR(o);
```

Comment:
listdir() should return a list, not an iterator.  See extmod/vfs.c for an implementation of listdir that uses ilistdir.